### PR TITLE
G-API: Parameterized render tests

### PR DESCRIPTION
**Pull request by TolyaTalamanov**

### CODE REVIEW

File: modules/gapi/test/common/gapi_render_tests.hpp [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    void Run(cv::Mat& out_mat, cv::Mat& ref_mat, bool isNV12)
+    {
+        if (isNV12) {
+            cv::gapi::wip::draw::BGR2NV12(out_mat, y, uv);
+            cv::gapi::wip::draw::render(y, uv, prims);
+            cv::cvtColorTwoPlane(y, uv, out_mat, cv::COLOR_YUV2BGR_NV12);
+
+            cv::gapi::wip::draw::BGR2NV12(ref_mat, y, uv);
```
Comment (andrey-golubev) [inline modules/gapi/test/common/gapi_render_tests.hpp]: I'd comment that this is done due to data loss when converting to NV12 and back, that's why `ref_mat` undergoes the same transformation

File: modules/gapi/test/cpu/gapi_render_tests_cpu.cpp [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+#include "../common/gapi_render_tests.hpp"
+
+namespace opencv_test
+{
+
+INSTANTIATE_TEST_CASE_P(RenderTextTestCPU, RenderTextTest,
+                        Combine(Values(cv::Size(1280, 720),
+                                       cv::Size(640, 480),
+                                       cv::Size(128, 128)),
+                                Values("text"),
```
Comment (andrey-golubev) [inline modules/gapi/test/cpu/gapi_render_tests_cpu.cpp]: just curious what if the text is empty? or "\0"?

File: modules/gapi/test/cpu/gapi_render_tests_cpu.cpp [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+                                Values(Points{Point(5, 30), Point(40, 70), Point(75, 110)}),
```
Comment (andrey-golubev) [inline modules/gapi/test/cpu/gapi_render_tests_cpu.cpp]: what if I pass `Point(-1, -1)`?

File: modules/gapi/test/cpu/gapi_render_tests_cpu.cpp [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+/* Thickness          */        Values(1),
+/* Line type          */        Values(LINE_8),
+/* Bottom left origin */        testing::Bool(),
+/* NV12 format or not */        testing::Bool()));
+
+INSTANTIATE_TEST_CASE_P(RenderRectTestCPU, RenderRectTest,
+                        Combine(Values(cv::Size(1280, 720),
+                                       cv::Size(640, 480),
+                                       cv::Size(128, 128)),
+                                Values(Rects{Rect(5, 30, 40, 50), Rect(40, 70, 40, 50), Rect(75, 110, 40, 50)}),
```
Comment (andrey-golubev) [inline modules/gapi/test/cpu/gapi_render_tests_cpu.cpp]: what if `Rect` has 0 or negative width/height?

File: modules/gapi/test/common/gapi_render_tests.hpp [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+        TestFunctional::initMatsRandU(type, size, type, initOut);
+
+        if (isNV12) {
+            cv::gapi::wip::draw::BGR2NV12(TestFunctional::out_mat_ocv, y, uv);
+            cv::cvtColorTwoPlane(y, uv, TestFunctional::out_mat_ocv, cv::COLOR_YUV2BGR_NV12);
+        }
+    }
+
+    void Run(cv::Mat& out_mat, cv::Mat& ref_mat, bool isNV12)
```
Comment (andrey-golubev) [inline modules/gapi/test/common/gapi_render_tests.hpp]: seems like you don't need `bool isNV12` since there's a `isNV12Format` member variable already

**Comment by TolyaTalamanov** [comment]
@dmatveev

File: modules/gapi/test/common/gapi_render_tests.hpp (hunk 5 above)
Comment (TolyaTalamanov) [reply modules/gapi/test/common/gapi_render_tests.hpp]: Fixed

File: modules/gapi/test/cpu/gapi_render_tests_cpu.cpp (hunk 2 above)
Comment (TolyaTalamanov) [reply modules/gapi/test/cpu/gapi_render_tests_cpu.cpp]: A blank text will be printed, this is an interesting case, but it increases the number of tests twice, I don't think it's worth it. What do you think ?

File: modules/gapi/test/common/gapi_render_tests.hpp (hunk 1 above)
Comment (TolyaTalamanov) [reply modules/gapi/test/common/gapi_render_tests.hpp]: You're right, added

**Comment by TolyaTalamanov** [comment]
@dmatveev

File: modules/gapi/test/cpu/gapi_render_tests_cpu.cpp (hunk 3 above)
Comment (TolyaTalamanov) [reply modules/gapi/test/cpu/gapi_render_tests_cpu.cpp]: Added

File: modules/gapi/test/cpu/gapi_render_tests_cpu.cpp (hunk 4 above)
Comment (TolyaTalamanov) [reply modules/gapi/test/cpu/gapi_render_tests_cpu.cpp]: Added

File: modules/gapi/test/cpu/gapi_render_tests_cpu.cpp (hunk 2 above)
Comment (andrey-golubev) [reply modules/gapi/test/cpu/gapi_render_tests_cpu.cpp]: I guess it's fine not to add it explicitly, just wondering whether you tried such stuff out to ensure there are no segfauls, unexpected behaviors, etc.

File: modules/gapi/test/common/gapi_render_tests.cpp [hunk 6]
```diff
@@ -0,0 +1,9 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+// Copyright (C) 2018 Intel Corporation
```
Comment (andrey-golubev) [inline modules/gapi/test/common/gapi_render_tests.cpp]: Let's update?
```suggestion
// Copyright (C) 2019 Intel Corporation
```

Comment (andrey-golubev) [reply modules/gapi/test/common/gapi_render_tests.cpp]: same in other places

File: modules/gapi/test/common/gapi_render_tests.hpp [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+    void Run()
+    {
+        if (isNV12Format) {
+            cv::gapi::wip::draw::BGR2NV12(out_mat_gapi, y, uv);
+            cv::gapi::wip::draw::render(y, uv, prims);
+            cv::cvtColorTwoPlane(y, uv, out_mat_gapi, cv::COLOR_YUV2BGR_NV12);
+
+            // NB: Also due to data loss
+            cv::gapi::wip::draw::BGR2NV12(out_mat_ocv, y, uv);
```
Comment (andrey-golubev) [inline modules/gapi/test/common/gapi_render_tests.hpp]: hmm, looks like you do this already in `Init()`, why do it again? btw, doesn't it mean you have "bigger" data loss for `out_mat_ocv` now?

File: modules/gapi/test/common/gapi_render_tests.hpp (hunk 1 above)
Comment (andrey-golubev) [reply modules/gapi/test/common/gapi_render_tests.hpp]: consider this resolved. for some reason I cannot resolve my discussions so feel free to resolve yourself

File: modules/gapi/test/cpu/gapi_render_tests_cpu.cpp (hunk 2 above)
Comment (andrey-golubev) [reply modules/gapi/test/cpu/gapi_render_tests_cpu.cpp]: consider resolved

File: modules/gapi/test/cpu/gapi_render_tests_cpu.cpp (hunk 3 above)
Comment (andrey-golubev) [reply modules/gapi/test/cpu/gapi_render_tests_cpu.cpp]: consider resolved

File: modules/gapi/test/cpu/gapi_render_tests_cpu.cpp (hunk 4 above)
Comment (andrey-golubev) [reply modules/gapi/test/cpu/gapi_render_tests_cpu.cpp]: consider resolved

File: modules/gapi/test/common/gapi_render_tests.hpp (hunk 5 above)
Comment (andrey-golubev) [reply modules/gapi/test/common/gapi_render_tests.hpp]: consider resolved

File: modules/gapi/test/common/gapi_render_tests.cpp (hunk 6 above)
Comment (dmatveev) [reply modules/gapi/test/common/gapi_render_tests.cpp]: I think that could be aligned generally with some annual copyrights update.

File: modules/gapi/test/common/gapi_render_tests.hpp [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+template<class T>
+class RenderWithParam : public TestWithParam<T>
+{
+protected:
+    void Init()
+    {
+        MatType type = CV_8UC3;
+        out_mat_ocv  = cv::Mat(sz, type, cv::Scalar(0));
+        out_mat_gapi = cv::Mat(sz, type, cv::Scalar(0));
```
Comment (dmatveev) [inline modules/gapi/test/common/gapi_render_tests.hpp]: Painting on a black image?
Probably painting on a white one (255) would be, hm, closer to the domain?

Comment (TolyaTalamanov) [reply modules/gapi/test/common/gapi_render_tests.hpp]: Fixed

File: modules/gapi/test/common/gapi_render_tests.hpp (hunk 7 above)
Comment (TolyaTalamanov) [reply modules/gapi/test/common/gapi_render_tests.hpp]: Let's discuss this locally

File: modules/gapi/test/common/gapi_render_tests.cpp (hunk 6 above)
Comment (andrgolubev) [reply modules/gapi/test/common/gapi_render_tests.cpp]: makes sense. consider resolved from my side

Comment (andrey-golubev) [reply modules/gapi/test/common/gapi_render_tests.cpp]: makes sense, consider resolved from my side

File: modules/gapi/test/common/gapi_render_tests.hpp (hunk 7 above)
Comment (andrey-golubev) [reply modules/gapi/test/common/gapi_render_tests.hpp]: if that's intended then I'm OK. consider this one resolved

**Comment by andrey-golubev** [comment]
@alalek I guess this one is merge-able, please merge

**Pull request merged**